src/BlockChyp/Client/TerminalRouteCache.cs

```diff
@@ -124,23 +120,65 @@ private TerminalRouteResponse GetOffline(string key, ApiCredentials rootCredenti
                 return null;
             }
 
-            using (var file = new StreamReader(cacheFile))
+            string rawContent = ReadFile(cacheFile);
+            try
             {
-                var raw = file.ReadToEnd();
-                try
-                {
-                    var result = JsonConvert.DeserializeObject<TerminalRouteResponse>(raw);
-                    result.TransientCredentials = Decrypt(result.TransientCredentials, rootCredentials);
+                var result = JsonConvert.DeserializeObject<TerminalRouteResponse>(rawContent);
+                result.TransientCredentials = Decrypt(result.TransientCredentials, rootCredentials);
+
+                return result;
+            }
+            catch (JsonException)
+            {
+                // File is invalid, remove it
+                DeleteFile(cacheFile);
+                return null;
+            }
+        }
 
-                    return result;
+        private string ReadFile(string path)
```

Comment:
Perhaps consider renaming these `*File` methods to `*OfflineCache` or `*PersistentCache`? Perhaps there's also a prefix or suffix we could append to the names in an effort to make it clear that these are in fact synchronized/locking operations. 

My first thought was to explicitly pass the `ReaderWriterLockSlim` value as a parameter to each of these functions. This would make it pretty obvious and also could be good for testing with a custom resource locking mechanism (although, not sure if the test suite will ever go that far).

Reply:
These are private methods attached to a cache class. If anyone ever needs to look at them (which is unlikely), it's evident what they do and why they exist. I think testing this functionality in isolation would add more maintenance overhead than value.

Reply:
We should first establish that anyone who has to maintain the code after you **will** have to look at them. This isn't unlikely at all but actually something to be expected, especially in an active open-source project (which I'm assuming is part of the goal here).

It is evident what they do in the event that you look at the definition, but it is not clear by the names, type signatures, or documentation that these methods use locking under-the-hood. i.e., they are misleading.

For example, if I were simply looking through the definition of `TerminalRouteCache.GetOffline`, it's actually **inevident** why the `ReadFile` method exists as a member of `TerminalRouteCache` (or any of the others, for that matter). One might even consider that they have to ensure that they acquire a lock before calling it. You would likely be forced to go and look at the definition in order to reason about its existence (unnecessary programmer overhead).

We can agree to disagree, though. Since they're private methods in a relatively small (_for now_) project, it's not a big deal.

Reply:
I'll agree on the testing front, though.

I don't think it's worth writing tests for this right now, but I wouldn't dismiss the importance of having an implementation with decent assurance of thread-safety. If this were to deadlock on a client's system that would be pretty lame, no?